lib/modules/manager/cargo/artifacts.ts

```diff
@@ -29,6 +31,46 @@ async function cargoUpdate(
   await exec(cmd, execOptions);
 }
 
+async function cargoUpdatePrecise(
+  manifestPath: string,
+  updatedDeps: Upgrade<Record<string, unknown>>[],
+  constraint: string | undefined
+): Promise<void> {
+  const cmds = updatedDeps
+    .filter((dep) => !!dep.lockedVersion)
+    .map(
+      (dep) =>
+        `cargo update --manifest-path ${quote(manifestPath)}` +
+        ` --package ${dep.packageName!}@${dep.lockedVersion}` +
+        (dep.newVersion ? ` --precise ${dep.newVersion}` : '')
```

Comment:
i don't think newVersion can be missing here. Renovate would filter those before calling manager